tree/treeplayer/inc/ROOT/TDFInterface.hxx

```diff
@@ -1182,7 +1182,10 @@ protected:
             TDirectory *subdir = nullptr;
             if (!dirnameInt.empty()) subdir = files[slot]->mkdir(dirnameInt.c_str());
             trees[slot] = new TTree(treenameInt.c_str(), treenameInt.c_str());
-            trees[slot]->SetDirectory(subdir ? subdir : files[slot].get());
+            {
+               R__LOCKGUARD(gROOTMutex);
```

Comment:
What global state does SetDirectory changes?

Reply:
The reference directory of the Tree?

Reply:
@pcanal you are right, looking again it's not a _global_ state, it's a shared state, sorry.
The lock fixes is, but I'm not sure why at this point.
Before adding this lockguard, this is the kind of stacktrace I got when running `roottest/root/dataframe/test_snapshotNFiles.C`:
```

===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================

Thread 4 (Thread 0x7fb273a97700 (LWP 26142)):
#0  0x00007fb27e34500a in __GI___waitpid (pid=26173, stat_loc=stat_loc
entry=0x7fb273a8f020, options=options
entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:29
#1  0x00007fb27e2cc07b in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:148
#2  0x00007fb2805e0402 in TUnixSystem::Exec (shellcmd=<optimized out>, this=0x55d42022f4f0) at /home/eguiraud/ROOT/root/core/unix/src/TUnixSystem.cxx:2118
#3  TUnixSystem::StackTrace (this=0x55d42022f4f0) at /home/eguiraud/ROOT/root/core/unix/src/TUnixSystem.cxx:2412
#4  0x00007fb2805e288c in TUnixSystem::DispatchSignals (this=0x55d42022f4f0, sig=kSigSegmentationViolation) at /home/eguiraud/ROOT/root/core/unix/src/TUnixSystem.cxx:3643
#5  <signal handler called>
#6  0x00007fb280562a9d in TList::FindLink (this=this
entry=0x55d4202351a0, obj=obj
entry=0x7fb268014e20, idx=
0x7fb273a9170c: 66) at /home/eguiraud/ROOT/root/core/cont/src/TList.cxx:545
#7  0x00007fb280562d7c in TList::Remove (this=this
entry=0x55d4202351a0, obj=obj
entry=0x7fb268014e20) at /home/eguiraud/ROOT/root/core/cont/src/TList.cxx:684
#8  0x00007fb2805602de in THashList::Remove (this=0x55d4202351a0, obj=0x7fb268014e20) at /home/eguiraud/ROOT/root/core/cont/src/THashList.cxx:289
#9  0x00007fb27f9b1a60 in TTree::SetDirectory (this=0x7fb268014e20, dir=0x55d422519250) at /home/eguiraud/ROOT/root/tree/tree/src/TTree.cxx:8442
#10 0x00007fb27530d773 in ?? ()
#11 0x0000000000000001 in ?? ()
#12 0x00007fb273a91788 in ?? ()
#13 0x0000000000000000 in ?? ()

Thread 3 (Thread 0x7fb273e98700 (LWP 26141)):
#0  0x00007fb27e34500a in __GI___waitpid (pid=26159, stat_loc=stat_loc
entry=0x7fb273e90020, options=options
entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:29
#1  0x00007fb27e2cc07b in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:148
#2  0x00007fb2805e0402 in TUnixSystem::Exec (shellcmd=<optimized out>, this=0x55d42022f4f0) at /home/eguiraud/ROOT/root/core/unix/src/TUnixSystem.cxx:2118
#3  TUnixSystem::StackTrace (this=0x55d42022f4f0) at /home/eguiraud/ROOT/root/core/unix/src/TUnixSystem.cxx:2412
#4  0x00007fb2805e288c in TUnixSystem::DispatchSignals (this=0x55d42022f4f0, sig=kSigSegmentationViolation) at /home/eguiraud/ROOT/root/core/unix/src/TUnixSystem.cxx:3643
#5  <signal handler called>
#6  0x00007fb280562a9d in TList::FindLink (this=this
entry=0x55d4202351a0, obj=obj
entry=0x7fb264013c80, idx=
0x7fb273e9270c: 66) at /home/eguiraud/ROOT/root/core/cont/src/TList.cxx:545
#7  0x00007fb280562d7c in TList::Remove (this=this
entry=0x55d4202351a0, obj=obj
entry=0x7fb264013c80) at /home/eguiraud/ROOT/root/core/cont/src/TList.cxx:684
#8  0x00007fb2805602de in THashList::Remove (this=0x55d4202351a0, obj=0x7fb264013c80) at /home/eguiraud/ROOT/root/core/cont/src/THashList.cxx:289
#9  0x00007fb27f9b1a60 in TTree::SetDirectory (this=0x7fb264013c80, dir=0x55d423555060) at /home/eguiraud/ROOT/root/tree/tree/src/TTree.cxx:8442
#10 0x00007fb27530d773 in ?? ()
#11 0x0000000000000001 in ?? ()
#12 0x00007fb273e92788 in ?? ()
#13 0x0000000000000000 in ?? ()

Thread 2 (Thread 0x7fb274699700 (LWP 26140)):
#0  pthread_cond_wait

GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
#1  0x00007fb27ee1850c in std::condition_variable::wait(std::unique_lock<std::mutex>&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x00007fb2800e06fb in std::condition_variable::wait<ROOT::Experimental::TBufferMerger::WriteOutputFile()::<lambda()> > (__p=..., __lock=..., this=0x7ffd9d64b5e8) at /usr/include/c++/6/condition_variable:99
#3  ROOT::Experimental::TBufferMerger::WriteOutputFile (this=0x7ffd9d64b570) at /home/eguiraud/ROOT/root/io/io/src/TBufferMerger.cxx:72
#4  0x00007fb27ee1de6f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007fb27e633494 in start_thread (arg=0x7fb274699700) at pthread_create.c:333
#6  0x00007fb27e375aff in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97

Thread 1 (Thread 0x7fb27c99dac0 (LWP 26114)):
#0  0x00007fb27e34500a in __GI___waitpid (pid=26145, stat_loc=stat_loc
entry=0x7ffd9d6485e0, options=options
entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:29
#1  0x00007fb27e2cc07b in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:148
#2  0x00007fb2805e0402 in TUnixSystem::Exec (shellcmd=<optimized out>, this=0x55d42022f4f0) at /home/eguiraud/ROOT/root/core/unix/src/TUnixSystem.cxx:2118
#3  TUnixSystem::StackTrace (this=0x55d42022f4f0) at /home/eguiraud/ROOT/root/core/unix/src/TUnixSystem.cxx:2412
#4  0x00007fb2805e288c in TUnixSystem::DispatchSignals (this=0x55d42022f4f0, sig=kSigSegmentationViolation) at /home/eguiraud/ROOT/root/core/unix/src/TUnixSystem.cxx:3643
#5  <signal handler called>
#6  0x0000000000000041 in ?? ()
#7  0x00007fb280562069 in TList::Clear (this=0x7fb268013ab0, option=<optimized out>) at /home/eguiraud/ROOT/root/core/cont/src/TList.cxx:391
#8  0x00007fb2805634ce in TList::~TList (this=0x7fb268013ab0, __in_chrg=<optimized out>) at /home/eguiraud/ROOT/root/core/cont/src/TList.cxx:91
#9  0x00007fb2805634f9 in TList::~TList (this=0x7fb268013ab0, __in_chrg=<optimized out>) at /home/eguiraud/ROOT/root/core/cont/src/TList.cxx:92
#10 0x00007fb280560bbc in THashTable::Remove (this=0x55d420235200, obj=<optimized out>) at /home/eguiraud/ROOT/root/core/cont/src/THashTable.cxx:323
#11 0x00007fb27f9b1a60 in TTree::SetDirectory (this=0x55d42384a8d0, dir=0x55d422438c40) at /home/eguiraud/ROOT/root/tree/tree/src/TTree.cxx:8442
#12 0x00007fb27530d773 in ?? ()
#13 0x0000000000000001 in ?? ()
#14 0x00007ffd9d64ad48 in ?? ()
#15 0x0000000000000000 in ?? ()
===========================================================


The lines below might hint at the cause of the crash.
You may get help by asking at the ROOT forum http://root.cern.ch/forum.
Only if you are really convinced it is a bug in ROOT then please submit a
report at http://root.cern.ch/bugs. Please post the ENTIRE stack trace
from above as an attachment in addition to anything else
that might help us fixing this issue.
===========================================================
#6  0x00007fb280562a9d in TList::FindLink (this=this
entry=0x55d4202351a0, obj=obj
entry=0x7fb268014e20, idx=
0x7fb273a9170c: 66) at /home/eguiraud/ROOT/root/core/cont/src/TList.cxx:545
#7  0x00007fb280562d7c in TList::Remove (this=this
entry=0x55d4202351a0, obj=obj
entry=0x7fb268014e20) at /home/eguiraud/ROOT/root/core/cont/src/TList.cxx:684
#8  0x00007fb2805602de in THashList::Remove (this=0x55d4202351a0, obj=0x7fb268014e20) at /home/eguiraud/ROOT/root/core/cont/src/THashList.cxx:289
#9  0x00007fb27f9b1a60 in TTree::SetDirectory (this=0x7fb268014e20, dir=0x55d422519250) at /home/eguiraud/ROOT/root/tree/tree/src/TTree.cxx:8442
#10 0x00007fb27530d773 in ?? ()
#11 0x0000000000000001 in ?? ()
#12 0x00007fb273a91788 in ?? ()
#13 0x0000000000000000 in ?? ()

===========================================================
```

Reply:
Different solution: we do a `cd` in the right directory right before constructing the tree instead of constructing the tree and then changing its directory. It is guaranteed that no-one else will modify the gDirectory in the middle.

Reply:
> instead of constructing the tree and then changing its directory.

Fair enough.  This was the problem.  The 'first' registration was (according to the stack trace) done in the same directory object from several threads (which is indeed not supported).  In the master, you could also avoid all of this by passing the target directory to the TTree constructor.

Reply:
> The 'first' registration was (according to the stack trace) done in the same directory object from several threads (which is indeed not supported).

ah, this should not happen, we try to leave the `gDirectories` unmodified by using `TContext`s. We might have missed a spot. Where do you see this in the stack?

>  In the master, you could also avoid all of this by passing the target directory to the TTree constructor.

we do :)

Reply:
> Where do you see this in the stack?
3 of the threads are in:
```
#8  0x00007fb2805602de in THashList::Remove (this=0x55d4202351a0, obj=0x7fb268014e20) at /home/eguiraud/ROOT/root/core/cont/src/THashList.cxx:289
#9  0x00007fb27f9b1a60 in TTree::SetDirectory (this=0x7fb268014e20, dir=0x55d422519250) at /home/eguiraud/ROOT/root/tree/tree/src/TTree.cxx:8442
```
Which the same value of THashList's this (0x55d4202351a0).

> we try to leave the gDirectories unmodified by using TContexts.

Which means it is probably still point to gROOT (which really we should also protect against that).

Reply:
wait..`gROOT` is shared among threads?

Reply:
gROOT point to a singleton.  and gDirectory most often default to point to it. [Note that we probably need to overload the TDirectory functionality of TROOT to make it thread safe.  This can be done now (as the 'hard' part is adding the functions overload) and later retrofitted to use the read-write lock.

Reply:
So `gDirectory` is a thread-local that is implicitly set and reset under our feet and implicitly picks up ownership of the objects we create if we are not careful, and by default all the thread local `gDirectories` point to a singleton, _but this is thread-unsafe_.

All of this is...very hard to deal with correctly, let alone reason about.